controller/src/index.ts

```
@@ -288,6 +292,54 @@ class Controller {
 
     return this.keychain.approvals(origin);
   }
+
+  async verifyMessageHash(
+    messageHash: number.BigNumberish,
+    signature: Signature
+  ) {
+    const isDeployed = !!this.account?.getClassHashAt(this.account.address, "latest");
+
+    if (isDeployed) {
+      const res = await this.account?.callContract({
+        contractAddress: this.account.address,
+        entrypoint: "executeOnPlugin",
+        calldata: [
+          CLASS_HASHES["0.0.1"].controller,
+          hash.getSelector("is_public_key"),
+          "0x1",
+          signature[0],
+        ]
+      });
+
+      const isRegistered = res?.result[0] === "0x1";
+      if (isRegistered) {
+        const keyPair = ec.getKeyPairFromPublicKey(signature[0]);
+        return ec.verify(keyPair, number.toBN(messageHash).toString(), signature);
+      } else {
+        // validate register txn and signature
+      }
+    } else {
+      const res = await client.request(AccountDocument, {
+        address: this.account?.address,
+      });
+
+      const account = res?.accounts?.edges?.[0]?.node;
+      if (!account) {
+        return false;
+      }
+
+      const pubKeyCbor = cbor.decodeAllSync(number.toBN(account.credential.publicKey).toBuffer())[0];
+      const x = number.toBN("0x" + pubKeyCbor.get(-2).toString("hex"));
+      const y = number.toBN("0x" + pubKeyCbor.get(-3).toString("hex"));
+      const { x: x0, y: x1, z: x2 } = split(x);
+      const { x: y0, y: y1, z: y2 } = split(y);
+      const address = computeAddress(account.id, {x0, x1, x2}, {y0, y1, y2}, signature[0]);
+      assert(address === this.account?.address, "invalid public key");
```

lets throw an error here